pkg/apply/prune/prune_test.go

```diff
@@ -183,155 +153,487 @@ var preventDelete = &unstructured.Unstructured{
+				t.Fatalf("expected %d prune events, got %d", len(tc.expectedEvents), len(actualEvents))
+			}
+			for _, expected := range tc.expectedEvents {
+				if !eventExists(actualEvents, expected) {
+					t.Errorf("expected event (%#v) not found in actual events", expected)
 				}
-				err := func() error {
-					defer close(eventChannel)
-					// Run the prune and validate.
-					return po.Prune(currentInventory, tc.currentObjs, populateObjectIds(tc.currentObjs), taskContext, Options{
-						DryRunStrategy: drs,
-					})
-				}()
+			}
+		})
+	}
+}
 
-				if err != nil {
-					t.Fatalf("Unexpected error during Prune(): %#v", err)
+func TestPruneWithErrors(t *testing.T) {
+	tests := map[string]struct {
+		pruneObjs      []*unstructured.Unstructured
+		destroy        bool
+		expectedEvents []event.Event
+	}{
+		"Prune delete failure": {
+			pruneObjs: []*unstructured.Unstructured{pdbDeleteFailure},
+			expectedEvents: []event.Event{
+				{
+					Type: event.PruneType,
+					PruneEvent: event.PruneEvent{
+						Identifier: object.UnstructuredToObjMeta(pdbDeleteFailure),
+						Error:      fmt.Errorf("expected delete error"),
+					},
+				},
+			},
+		},
+		"Destroy delete failure": {
+			pruneObjs: []*unstructured.Unstructured{pdbDeleteFailure},
+			destroy:   true,
+			expectedEvents: []event.Event{
+				{
+					Type: event.DeleteType,
+					DeleteEvent: event.DeleteEvent{
+						Identifier: object.UnstructuredToObjMeta(pdbDeleteFailure),
+						Error:      fmt.Errorf("expected delete error"),
+					},
+				},
+			},
+		},
+	}
+	for name, tc := range tests {
+		t.Run(name, func(t *testing.T) {
+			po := NewPruneOptions()
+			po.Destroy = tc.destroy
+			pruneIds := object.UnstructuredsToObjMetas(tc.pruneObjs)
+			fakeInvClient := inventory.NewFakeInventoryClient(pruneIds)
+			po.InvClient = fakeInvClient
+			currentInventory := createInventoryInfo(tc.pruneObjs...)
+			// Set up the fake dynamic client to recognize all objects, and the RESTMapper.
+			po.Client = &fakeDynamicFailureClient{dynamic: fake.NewSimpleDynamicClient(scheme.Scheme,
+				namespace, pdb, role)}
+			po.Mapper = testrestmapper.TestOnlyStaticRESTMapper(scheme.Scheme,
+				scheme.Scheme.PrioritizedVersionsAllGroups()...)
+			// The event channel can not block; make sure its bigger than all
+			// the events that can be put on it.
+			eventChannel := make(chan event.Event, len(tc.pruneObjs))
+			taskContext := taskrunner.NewTaskContext(eventChannel)
+			err := func() error {
+				defer close(eventChannel)
+				// Run the prune and validate.
+				return po.Prune(currentInventory, tc.pruneObjs, sets.NewString(), taskContext, defaultOptions)
+			}()
+			if err != nil {
+				t.Fatalf("Unexpected error during Prune(): %#v", err)
+			}
+			var actualEvents []event.Event
+			for e := range eventChannel {
+				actualEvents = append(actualEvents, e)
+			}
+			// Validate the expected/actual error events
+			if len(tc.expectedEvents) != len(actualEvents) {
```

Comment:
Can we reuse the validation logic for events?
https://github.com/kubernetes-sigs/cli-utils/blob/1c8e8871efb7082a83d88250157373d4a29effd2/pkg/testutil/events.go#L53

We use this code in the e2e tests and the applier tests.

Reply:
Updated prune unit tests to use the `testutil.VerifyEvents` function